src/artifact-downloader.ts

```diff
@@ -0,0 +1,357 @@
+ * @param artifactName The name of the artifact to fetch.
+ * @param timeoutMs Timeout in milliseconds (default: 10000)
+ * @returns The full file as a Uint8Array.
+ */
+async function fetchFromIPFS (
+  rootCid: string,
+  artifactVariantString: string,
+  artifactName: ArtifactName,
+  timeoutMs: number = 10000
+): Promise<Uint8Array> {
+  // TODO: Make cache implementation and return path to cached file if it exists
+
+  await initHelia()
+
+  if (!fs) throw new Error('Helia UnixFS not initialized')
+
+  const cid = CID.parse(rootCid)
+  const path = getPathForArtifactName(artifactName, artifactVariantString)
+
+  dbg(`Fetching from IPFS CID: ${cid.toString()}${`/${path}`}`)
+
+  // Create timeout promise
+  const timeoutPromise = new Promise<never>((_resolve, reject) => {
+    setTimeout(() => reject(new Error(`IPFS fetch timeout after ${timeoutMs}ms`)), timeoutMs)
+  })
+
+  // Fetch with timeout
+  const fetchPromise = (async () => {
+    try {
+      const contents = fs.cat(cid, { path })
+
+      const chunks: Uint8Array[] = []
+      for await (const chunk of contents) {
+        chunks.push(chunk)
+      }
+
+      const totalLength = chunks.reduce((sum, chunk) => sum + chunk.length, 0)
+      const result = new Uint8Array(totalLength)
+      let offset = 0
+      for (const chunk of chunks) {
+        result.set(chunk, offset)
+        offset += chunk.length
+      }
+
+      // Decompress the artifact if needed
+      const decompressedResult = decompressIfNeeded(result, artifactName)
+
+      const isValid = await validateArtifactDownload(
+        decompressedResult,
+        artifactName,
+        artifactVariantString
+      )
+
+      if (isValid) {
+      // TODO: Store it and return path where it was stored.
+        return result
+      } else {
+        throw new Error(
+        `Invalid hash for artifact download: ${artifactName} for ${artifactVariantString}.`
+        )
+      }
+    } catch (error) {
+      throw new Error(`IPFS fetch failed for ${artifactName} (${artifactVariantString}): ${error instanceof Error ? error.message : String(error)}`)
+    }
+  })()
+
+  return await Promise.race([fetchPromise, timeoutPromise])
+}
+
+/**
+ * Fetches from IPFS with HTTP gateway fallback for reliability
+ * @param rootCid The root CID string (e.g., 'QmeBrG7pii1qTqsn7rusvDiqXopHPjCT9gR4PsmW7wXqZq')
+ * @param artifactVariantString The string representing the artifact variant (e.g., "2x16").
+ * @param artifactName The name of the artifact to fetch.
+ * @returns The full file as a Uint8Array
+ */
+async function fetchFromIPFSWithFallback (
+  rootCid: string,
+  artifactVariantString: string,
+  artifactName: ArtifactName
+): Promise<Uint8Array> {
+  // Try fetching from IPFS first, with a fallback to HTTP gateway if it fails
+  try {
+    console.log('trying ipfs')
+    // Try IPFS first with 10s timeout
+    return await fetchFromIPFS(rootCid, artifactVariantString, artifactName, 10000)
```

Comment:
Will add proper error handling so only axios is reverted to ONLY if the IPFS call fails due to a network failure / slow response

Reply:
add todo